File: day_3/part_2.py
def solve(inp):
    rows = []
    with open(inp, "r") as file:
        lines = file.readlines()
        for l in lines:
            rows.append(list(map(lambda x: int(x), l.replace("\n", ""))))

    res = 0
    for row in rows:
        search = len(row) - 12
        last_indx = -1
        arr = []
        for _ in range(12):
            add = (0, -1)
            for j in range(last_indx + 1, last_indx + search + 1):
                add = max(add, (row[j], j), key=lambda x: (x[0], -x[1]))
            arr.append(str(add[0]))
            last_indx = add[1]
            search = len(row) - last_indx - 12 + len(arr)
        res += int("".join(arr))

    return res


# print(solve("example.txt"))
print(solve("input.txt"))

File: day_3/part_2.cpp
#include <iostream>
#include <fstream>
#include <string> 

using namespace std;

vector<vector<int>> create_rows(){
    ifstream inputFile("input.txt");
    vector<vector<int>> rows;
    string line;
    while( !inputFile.eof() ) {
        getline( inputFile, line );
        vector<int> nums;
        for (char n: line) {
            nums.push_back(n - '0');
        }
        rows.push_back(nums);
    }
    
    return rows;
}

int main() {
    vector<vector<int>> rows = create_rows();
    long long res = 0;
    for(vector<int> row : rows) {
        int search = row.size() - 12;
        int last_indx = -1;
        vector<int> arr;
        for (int i = 0; i < 12; i++) {
            int add[2] = {0, -1};
            for (int j = last_indx + 1; j < last_indx + search + 1; j++) {
                if(row[j] > add[0]) {
                    add[0] = row[j];
                    add[1] = j;
                }
            }
            arr.push_back(add[0]);
            last_indx = add[1];
            search = row.size() - last_indx - 12 + arr.size();
        }
        string s;
        for (int d : arr) s += to_string(d);
        long long number = stoll(s); 
        res += number;
    }
    cout << res;
    return res;
};



